src/rules/noInvalidThisRule.ts

```diff
@@ -96,3 +101,7 @@ function walk(ctx: Lint.WalkContext<boolean>): void {
         ts.forEachChild(node, cb);
     });
 }
+
+function hasContextualThis(node: ts.FunctionDeclaration): boolean {
```

Comment:
please rename to `hasThisParameter` (or completely inline it)
This has nothing to do with **contextual** `this`. It's about explicitly declared `this` parameters.

Contextual `this` looks as follows:

```ts
declare function callWithThis(fn: (this: Array<any>) => void): void;

callWithThis(function() {
    console.log(this.length); // this is allowed, because `this` is contextually typed to be an array
});
```
To detect that you would need the TypeChecker. I don't think that's worth it, because TypeScript provides the option `--noImplicitThis` which makes this rule obsolete.